File: java_src/src/test/java/com/google/crypto/tink/keyderivation/KeyDerivationKeyTemplatesTest.java
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.keyderivation;

import static com.google.common.truth.Truth.assertThat;
import static com.google.crypto.tink.internal.Util.UTF_8;
import static org.junit.Assert.assertThrows;

import com.google.crypto.tink.Aead;
import com.google.crypto.tink.KeyTemplate;
import com.google.crypto.tink.KeyTemplate.OutputPrefixType;
import com.google.crypto.tink.KeyTemplates;
import com.google.crypto.tink.KeysetHandle;
import com.google.crypto.tink.aead.AeadConfig;
import com.google.crypto.tink.prf.PrfConfig;
import com.google.crypto.tink.testing.TestUtil;
import java.security.GeneralSecurityException;
import java.util.ArrayList;
import java.util.List;
import org.junit.Assume;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public final class KeyDerivationKeyTemplatesTest {

  @BeforeClass
  public static void setUp() throws Exception {
    KeyDerivationConfig.register();
    PrfConfig.register();
    AeadConfig.register();
  }

  @Test
  public void createPrfBasedKeyTemplate_succeeds() throws Exception {
    Assume.assumeFalse(TestUtil.isAndroid()); // Some Android versions don't support AES-GCM.

    KeyTemplate prfTemplate = KeyTemplates.get("HKDF_SHA256");
    KeyTemplate aeadTemplate = KeyTemplates.get("AES256_GCM");

    List<OutputPrefixType> outputPrefixTypes = new ArrayList<>();
    outputPrefixTypes.add(OutputPrefixType.TINK);
    outputPrefixTypes.add(OutputPrefixType.LEGACY);
    outputPrefixTypes.add(OutputPrefixType.RAW);
    outputPrefixTypes.add(OutputPrefixType.CRUNCHY);

    for (OutputPrefixType outputPrefixType : outputPrefixTypes) {
      KeyTemplate derivedTemplate =
          KeyTemplate.create(aeadTemplate.getTypeUrl(), aeadTemplate.getValue(), outputPrefixType);
      KeyTemplate prfBasedTemplate =
          KeyDerivationKeyTemplates.createPrfBasedKeyTemplate(prfTemplate, derivedTemplate);

      KeysetHandle handle = KeysetHandle.generateNew(prfBasedTemplate);
      KeysetDeriver deriver = handle.getPrimitive(KeysetDeriver.class);
      KeysetHandle derivedHandle = deriver.deriveKeyset("salty".getBytes(UTF_8));
      assertThat(derivedHandle.getKeysetInfo().getKeyInfoCount()).isEqualTo(1);
      assertThat(derivedHandle.getKeysetInfo().getKeyInfo(0).getOutputPrefixType().toString())
          .isEqualTo(outputPrefixType.toString());

      // Use derivedHandle, which contains an AES256_GCM key.
      Aead aead = derivedHandle.getPrimitive(Aead.class);
      byte[] plaintext = "plaintext".getBytes(UTF_8);
      byte[] associatedData = "associatedData".getBytes(UTF_8);
      byte[] ciphertext = aead.encrypt(plaintext, associatedData);
      assertThat(aead.decrypt(ciphertext, associatedData)).isEqualTo(plaintext);
    }
  }

  @Test
  public void createPrfBasedKeyTemplate_failsForNotPrf() throws Exception {
    assertThrows(
        GeneralSecurityException.class,
        () ->
            KeyDerivationKeyTemplates.createPrfBasedKeyTemplate(
                KeyTemplates.get("AES256_GCM"), KeyTemplates.get("AES256_GCM")));
  }

  @Test
  public void createPrfBasedKeyTemplate_failsForNotDerivableKeyType() throws Exception {
    KeyTemplate derivedTemplate =
        KeyDerivationKeyTemplates.createPrfBasedKeyTemplate(
            KeyTemplates.get("HKDF_SHA256"), KeyTemplates.get("AES256_GCM"));
    assertThrows(
        GeneralSecurityException.class,
        () ->
            KeyDerivationKeyTemplates.createPrfBasedKeyTemplate(
                KeyTemplates.get("HKDF_SHA256"), derivedTemplate));
  }
}

File: cc/keyderivation/key_derivation_key_templates_test.cc
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/keyderivation/key_derivation_key_templates.h"

#include <vector>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/memory/memory.h"
#include "absl/status/status.h"
#include "tink/aead/aead_key_templates.h"
#include "tink/aead/aes_gcm_key_manager.h"
#include "tink/keyderivation/internal/prf_based_deriver_key_manager.h"
#include "tink/keyderivation/keyset_deriver_wrapper.h"
#include "tink/prf/hkdf_prf_key_manager.h"
#include "tink/prf/prf_key_templates.h"
#include "tink/registry.h"
#include "tink/util/statusor.h"
#include "tink/util/test_matchers.h"
#include "proto/prf_based_deriver.pb.h"
#include "proto/tink.pb.h"

namespace crypto {
namespace tink {
namespace internal {
namespace {

using ::crypto::tink::test::IsOk;
using ::crypto::tink::test::StatusIs;
using ::google::crypto::tink::KeyTemplate;
using ::google::crypto::tink::OutputPrefixType;
using ::google::crypto::tink::PrfBasedDeriverKeyFormat;
using ::testing::Eq;
using ::testing::Not;

class KeyDerivationKeyTemplatesTest : public ::testing::Test {
 protected:
  void TearDown() override { Registry::Reset(); }
};

TEST_F(KeyDerivationKeyTemplatesTest, CreatePrfBasedKeyTemplate) {
  ASSERT_THAT(Registry::RegisterPrimitiveWrapper(
                  absl::make_unique<KeysetDeriverWrapper>()),
              IsOk());
  ASSERT_THAT(Registry::RegisterKeyTypeManager(
                  absl::make_unique<internal::PrfBasedDeriverKeyManager>(),
                  /*new_key_allowed=*/true),
              IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<HkdfPrfKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<AesGcmKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());

  std::vector<OutputPrefixType> output_prefix_types = {
      OutputPrefixType::RAW, OutputPrefixType::TINK, OutputPrefixType::LEGACY};
  for (OutputPrefixType output_prefix_type : output_prefix_types) {
    KeyTemplate derived_key_template = AeadKeyTemplates::Aes256Gcm();
    derived_key_template.set_output_prefix_type(output_prefix_type);
    util::StatusOr<KeyTemplate> key_template =
        KeyDerivationKeyTemplates::CreatePrfBasedKeyTemplate(
            PrfKeyTemplates::HkdfSha256(), derived_key_template);

    ASSERT_THAT(key_template, IsOk());
    EXPECT_THAT(
        key_template->type_url(),
        Eq("type.googleapis.com/google.crypto.tink.PrfBasedDeriverKey"));
    EXPECT_THAT(key_template->type_url(),
                Eq(internal::PrfBasedDeriverKeyManager().get_key_type()));
    EXPECT_THAT(key_template->output_prefix_type(), Eq(output_prefix_type));

    PrfBasedDeriverKeyFormat key_format;
    EXPECT_TRUE(key_format.ParseFromString(key_template->value()));
    EXPECT_THAT(
        internal::PrfBasedDeriverKeyManager().ValidateKeyFormat(key_format),
        IsOk());
  }
}

TEST_F(KeyDerivationKeyTemplatesTest, CreatePrfBasedKeyTemplateInvalidPrfKey) {
  ASSERT_THAT(Registry::RegisterPrimitiveWrapper(
                  absl::make_unique<KeysetDeriverWrapper>()),
              IsOk());
  ASSERT_THAT(Registry::RegisterKeyTypeManager(
                  absl::make_unique<internal::PrfBasedDeriverKeyManager>(),
                  /*new_key_allowed=*/true),
              IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<HkdfPrfKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<AesGcmKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());

  EXPECT_THAT(KeyDerivationKeyTemplates::CreatePrfBasedKeyTemplate(
                  AeadKeyTemplates::Aes256Gcm(), AeadKeyTemplates::Aes256Gcm())
                  .status(),
              StatusIs(absl::StatusCode::kNotFound));
}

TEST_F(KeyDerivationKeyTemplatesTest,
       CreatePrfBasedKeyTemplateInvalidDerivedKeyTemplate) {
  ASSERT_THAT(Registry::RegisterPrimitiveWrapper(
                  absl::make_unique<KeysetDeriverWrapper>()),
              IsOk());
  ASSERT_THAT(Registry::RegisterKeyTypeManager(
                  absl::make_unique<internal::PrfBasedDeriverKeyManager>(),
                  /*new_key_allowed=*/true),
              IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<HkdfPrfKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<AesGcmKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());

  util::StatusOr<KeyTemplate> derived_key_template =
      KeyDerivationKeyTemplates::CreatePrfBasedKeyTemplate(
          PrfKeyTemplates::HkdfSha256(), AeadKeyTemplates::Aes256Gcm());
  ASSERT_THAT(derived_key_template, IsOk());
  EXPECT_THAT(KeyDerivationKeyTemplates::CreatePrfBasedKeyTemplate(
                  PrfKeyTemplates::HkdfSha256(), *derived_key_template)
                  .status(),
              StatusIs(absl::StatusCode::kNotFound));
}

TEST_F(KeyDerivationKeyTemplatesTest,
       CreatePrfBasedKeyTemplateNoPrfBasedDeriverKeyManager) {
  ASSERT_THAT(Registry::RegisterPrimitiveWrapper(
                  absl::make_unique<KeysetDeriverWrapper>()),
              IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<HkdfPrfKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<AesGcmKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());

  EXPECT_THAT(KeyDerivationKeyTemplates::CreatePrfBasedKeyTemplate(
                  PrfKeyTemplates::HkdfSha256(), AeadKeyTemplates::Aes256Gcm()),
              Not(IsOk()));
}

TEST_F(KeyDerivationKeyTemplatesTest,
       CreatePrfBasedKeyTemplateNoHkdfPrfKeyManager) {
  ASSERT_THAT(Registry::RegisterPrimitiveWrapper(
                  absl::make_unique<KeysetDeriverWrapper>()),
              IsOk());
  ASSERT_THAT(Registry::RegisterKeyTypeManager(
                  absl::make_unique<internal::PrfBasedDeriverKeyManager>(),
                  /*new_key_allowed=*/true),
              IsOk());
  ASSERT_THAT(
      Registry::RegisterKeyTypeManager(absl::make_unique<AesGcmKeyManager>(),
                                       /*new_key_allowed=*/true),
      IsOk());

  EXPECT_THAT(KeyDerivationKeyTemplates::CreatePrfBasedKeyTemplate(
                  PrfKeyTemplates::HkdfSha256(), AeadKeyTemplates::Aes256Gcm()),
              Not(IsOk()));
}

}  // namespace
}  // namespace internal
}  // namespace tink
}  // namespace crypto
